[Bug] Runtime#unsafeRunToFuture is broken in RC-18

`Runtime#unsafeRunToFuture` is broken in RC-18 release as part of interruption rework in https://github.com/zio/zio/commit/681063606b157b7f98f51c664fc7e7a4c1a82918. I did a full bisect of the recent commits and that one was confirmed to be the cause.

Here's a quick reproducer:

```
import java.time.{LocalDate, LocalDateTime}

import zio.clock.Clock
import zio.console.{Console, putStrLn}
import zio.duration._
import zio.{Schedule, Task, ZIO}

object BrokenUnsafeRunToFutureReproducer extends zio.App {
  override def run(args: List[String]): ZIO[zio.ZEnv, Nothing, Int] = {
    val effect = ZIO.effectSuspendTotal(putStrLn(s"Current time is ${LocalDateTime.now}"))
      .delay(1.second)

    val toFutureAndBackToEffect = for {
      rt <- ZIO.runtime[Console with Clock]
      _ <- Task.fromFuture { _ =>
        rt.unsafeRunToFuture(effect)
      }
    } yield ()

    toFutureAndBackToEffect
      .repeat(Schedule.fixed(1.second))
      .orDie
      .as(0)
  }
}
```

Which will terminate with:

```
Fiber failed.
An unchecked error was produced.
java.util.concurrent.ExecutionException: Boxed Exception
	at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolve(Promise.scala:99)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:242)
	at scala.concurrent.Promise.complete(Promise.scala:57)
	at scala.concurrent.Promise.complete$(Promise.scala:56)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:104)
	at scala.concurrent.Promise.failure(Promise.scala:109)
	at scala.concurrent.Promise.failure$(Promise.scala:109)
	at scala.concurrent.impl.Promise$DefaultPromise.failure(Promise.scala:104)
	at zio.Fiber.$anonfun$toFutureWith$2(Fiber.scala:297)
	at zio.internal.FiberContext.evaluateNow(FiberContext.scala:383)
	at zio.internal.FiberContext.$anonfun$evaluateLater$1(FiberContext.scala:679)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.InterruptedException: Interrupted by fibers: #15
	at zio.Cause.$anonfun$squashWith$1(Cause.scala:348)
	at scala.Option.orElse(Option.scala:477)
	at zio.Cause.squashWith(Cause.scala:345)
	at zio.Cause.squashWith$(Cause.scala:343)
	at zio.Cause$Internal$Interrupt.squashWith(Cause.scala:648)
	at zio.Cause.squashWithTrace(Cause.scala:370)
	at zio.Cause.squashWithTrace$(Cause.scala:369)
	at zio.Cause$Internal$Interrupt.squashWithTrace(Cause.scala:648)
	... 6 more
Caused by: zio.Cause$FiberTrace: Fiber failed.
An interrupt was produced by #15.
No ZIO Trace available.
```

It's somewhat non-deterministic (hence the `repeat`).

--- a/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
@@ -1,56 +1,83 @@
 package zio
 
+import zio.clock.Clock
+import zio.console._
+import zio.duration._
 import zio.test.Assertion._
 import zio.test.TestAspect._
 import zio.test._
+import zio.test.environment._
 
 object CancelableFutureSpec extends ZIOBaseSpec {
+  def roundtrip[R, A](zio: RIO[R, A]): RIO[R, A] =
+    for {
+      future <- zio.toFuture
+      a      <- RIO.fromFuture(_ => future)
+    } yield a
 
-  def spec = suite("CancelableFutureSpec")(
-    testM("interrupts the underlying task on cancel 2") {
-      for {
-        p  <- Promise.make[Nothing, Unit]
-        p2 <- Promise.make[Nothing, Int]
-        f <- (p.succeed(()) *> IO.never)
-              .onInterrupt(p2.succeed(42))
-              .fork
-        _    <- p.await
-        _    <- f.interrupt
-        test <- p2.await
-      } yield assert(test)(equalTo(42))
-    } @@ nonFlaky,
-    testM("interrupts the underlying task on cancel") {
-      for {
-        p  <- Promise.make[Nothing, Unit]
-        p2 <- Promise.make[Nothing, Int]
-        f <- (p.succeed(()) *> IO.never)
-              .onInterrupt(p2.succeed(42))
-              .toFuture
-        _    <- p.await
-        _    <- ZIO.fromFuture(_ => f.cancel())
-        test <- p2.await
-      } yield assert(test)(equalTo(42))
-    } @@ nonFlaky,
-    testM("cancel returns the exit reason") {
-      val t = new Exception("test")
+  def spec =
+    suite("CancelableFutureSpec")(
+      testM("auto-kill regression") {
+        val effect = ZIO.unit.delay(1.millisecond)
 
-      for {
-        p1 <- Promise.make[Nothing, Unit]
-        p2 <- Promise.make[Nothing, Unit]
-        f1 <- (ZIO.succeedNow(42) <* p1.succeed(())).toFuture
-        f2 <- ZIO.failNow(t).onError(_ => p2.succeed(())).toFuture
-        _  <- p1.await *> p2.await
-        e1 <- ZIO.fromFuture(_ => f1.cancel())
-        e2 <- ZIO.fromFuture(_ => f2.cancel())
-      } yield assert(e1.succeeded)(isTrue) && assert(e2.succeeded)(isFalse)
-    },
-    testM("is a scala.concurrent.Future") {
-      for {
-        f <- ZIO(42).toFuture
-        v <- ZIO.fromFuture(_ => f)
-      } yield {
-        assert(v)(equalTo(42))
+        val roundtrip = for {
+          rt <- ZIO.runtime[Console with Clock]
+          _  <- Task.fromFuture(_ => rt.unsafeRunToFuture(effect))
+        } yield ()
+
+        val result = roundtrip.orDie.as(0)
+
+        assertM(Live.live(result))(equalTo(0))
+      } @@ tag("supervision", "regression") @@ nonFlaky,
+      testM("roundtrip preserves interruptibility") {
+        for {
+          latch <- Promise.make[Nothing, Int]
+          fiber <- roundtrip(ZIO.infinity.onInterrupt(latch.succeed(42))).fork
+          _     <- fiber.interrupt
+          value <- latch.await
+        } yield assert(value)(equalTo(42))
+      },
+      testM("survives roundtrip without being auto-killed") {
+        val exception = new Exception("Uh oh")
+        val value     = 42
+
+        for {
+          failure <- roundtrip(ZIO.fail(exception)).either
+          success <- roundtrip(ZIO.succeed(value)).either
+        } yield assert(failure)(isLeft(equalTo(exception))) && assert(success)(isRight(equalTo(value)))
+      } @@ tag("supervision") @@ nonFlaky,
+      testM("interrupts the underlying task on cancel") {
+        for {
+          p  <- Promise.make[Nothing, Unit]
+          p2 <- Promise.make[Nothing, Int]
+          f <- (p.succeed(()) *> IO.never)
+                .onInterrupt(p2.succeed(42))
+                .toFuture
+          _    <- p.await
+          _    <- ZIO.fromFuture(_ => f.cancel())
+          test <- p2.await
+        } yield assert(test)(equalTo(42))
+      } @@ nonFlaky,
+      testM("cancel returns the exit reason") {
+        val t = new Exception("test")
+
+        for {
+          p1 <- Promise.make[Nothing, Unit]
+          p2 <- Promise.make[Nothing, Unit]
+          f1 <- (ZIO.succeedNow(42) <* p1.succeed(())).toFuture
+          f2 <- ZIO.failNow(t).onError(_ => p2.succeed(())).toFuture
+          _  <- p1.await *> p2.await
+          e1 <- ZIO.fromFuture(_ => f1.cancel())
+          e2 <- ZIO.fromFuture(_ => f2.cancel())
+        } yield assert(e1.succeeded)(isTrue) && assert(e2.succeeded)(isFalse)
+      },
+      testM("is a scala.concurrent.Future") {
+        for {
+          f <- ZIO(42).toFuture
+          v <- ZIO.fromFuture(_ => f)
+        } yield {
+          assert(v)(equalTo(42))
+        }
       }
-    }
-  )
+    ) @@ tag("interop", "future")
 }

--- a/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
@@ -28,7 +28,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
         val result = roundtrip.orDie.as(0)
 
         assertM(Live.live(result))(equalTo(0))
-      } @@ tag("supervision", "regression") @@ nonFlaky,
+      } @@ jvm(nonFlaky) @@ tag("supervision", "regression"),
       testM("auto-kill regression 2") {
         val effect = clock.currentDateTime.map(_.toString()).delay(10.millisecond)
 
@@ -40,7 +40,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
         val result = roundtrip.orDie.forever
 
         assertM(Live.live(result.timeout(1.seconds)))(isNone)
-      } @@ tag("supervision", "regression"),
+      } @@ jvmOnly @@ tag("supervision", "regression"),
       testM("roundtrip preserves interruptibility") {
         for {
           start <- Promise.make[Nothing, Unit]

--- a/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
@@ -28,7 +28,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
         val result = roundtrip.orDie.as(0)
 
         assertM(Live.live(result))(equalTo(0))
-      } @@ jvmOnly @@ nonFlaky @@ tag("supervision", "regression"),
+      } @@ nonFlaky @@ tag("supervision", "regression"),
       testM("auto-kill regression 2") {
         val effect = clock.currentDateTime.map(_.toString()).delay(10.millisecond)
 
@@ -40,7 +40,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
         val result = roundtrip.orDie.forever
 
         assertM(Live.live(result.timeout(1.seconds)))(isNone)
-      } @@ jvmOnly @@ tag("supervision", "regression"),
+      } @@ tag("supervision", "regression"),
       testM("roundtrip preserves interruptibility") {
         for {
           start <- Promise.make[Nothing, Unit]
@@ -50,7 +50,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
           _     <- fiber.interrupt
           value <- end.await
         } yield assert(value)(equalTo(42))
-      } @@ jvmOnly @@ nonFlaky,
+      } @@ nonFlaky,
       testM("survives roundtrip without being auto-killed") {
         val exception = new Exception("Uh oh")
         val value     = 42
@@ -59,7 +59,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
           failure <- roundtrip(ZIO.fail(exception)).either
           success <- roundtrip(ZIO.succeed(value)).either
         } yield assert(failure)(isLeft(equalTo(exception))) && assert(success)(isRight(equalTo(value)))
-      } @@ jvmOnly @@ tag("supervision") @@ nonFlaky,
+      } @@ tag("supervision") @@ nonFlaky,
       testM("interrupts the underlying task on cancel") {
         for {
           p  <- Promise.make[Nothing, Unit]

--- a/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
@@ -31,12 +31,14 @@ object CancelableFutureSpec extends ZIOBaseSpec {
       } @@ tag("supervision", "regression") @@ nonFlaky,
       testM("roundtrip preserves interruptibility") {
         for {
-          latch <- Promise.make[Nothing, Int]
-          fiber <- roundtrip(ZIO.infinity.onInterrupt(latch.succeed(42))).fork
+          start <- Promise.make[Nothing, Unit]
+          end   <- Promise.make[Nothing, Int]
+          fiber <- roundtrip((start.succeed(()) *> ZIO.infinity).onInterrupt(end.succeed(42))).fork
+          _     <- start.await
           _     <- fiber.interrupt
-          value <- latch.await
+          value <- end.await
         } yield assert(value)(equalTo(42))
-      },
+      } @@ nonFlaky,
       testM("survives roundtrip without being auto-killed") {
         val exception = new Exception("Uh oh")
         val value     = 42

--- a/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
@@ -29,6 +29,18 @@ object CancelableFutureSpec extends ZIOBaseSpec {
 
         assertM(Live.live(result))(equalTo(0))
       } @@ tag("supervision", "regression") @@ nonFlaky,
+      testM("auto-kill regression 2") {
+        val effect = clock.currentDateTime.map(_.toString()).delay(10.millisecond)
+
+        val roundtrip = for {
+          rt <- ZIO.runtime[Console with Clock]
+          _  <- Task.fromFuture(_ => rt.unsafeRunToFuture(effect))
+        } yield ()
+
+        val result = roundtrip.orDie.forever
+
+        assertM(Live.live(result.timeout(1.seconds)))(isNone)
+      } @@ tag("supervision", "regression"),
       testM("roundtrip preserves interruptibility") {
         for {
           start <- Promise.make[Nothing, Unit]

--- a/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/CancelableFutureSpec.scala
@@ -28,7 +28,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
         val result = roundtrip.orDie.as(0)
 
         assertM(Live.live(result))(equalTo(0))
-      } @@ jvm(nonFlaky) @@ tag("supervision", "regression"),
+      } @@ jvmOnly @@ nonFlaky @@ tag("supervision", "regression"),
       testM("auto-kill regression 2") {
         val effect = clock.currentDateTime.map(_.toString()).delay(10.millisecond)
 
@@ -50,7 +50,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
           _     <- fiber.interrupt
           value <- end.await
         } yield assert(value)(equalTo(42))
-      } @@ nonFlaky,
+      } @@ jvmOnly @@ nonFlaky,
       testM("survives roundtrip without being auto-killed") {
         val exception = new Exception("Uh oh")
         val value     = 42
@@ -59,7 +59,7 @@ object CancelableFutureSpec extends ZIOBaseSpec {
           failure <- roundtrip(ZIO.fail(exception)).either
           success <- roundtrip(ZIO.succeed(value)).either
         } yield assert(failure)(isLeft(equalTo(exception))) && assert(success)(isRight(equalTo(value)))
-      } @@ tag("supervision") @@ nonFlaky,
+      } @@ jvmOnly @@ tag("supervision") @@ nonFlaky,
       testM("interrupts the underlying task on cancel") {
         for {
           p  <- Promise.make[Nothing, Unit]

--- a/core-tests/shared/src/test/scala/zio/ZIOSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/ZIOSpec.scala
@@ -1767,7 +1767,12 @@ object ZIOSpec extends ZIOBaseSpec {
       },
       testM("timeout a long computation") {
         val io = (clock.sleep(5.seconds) *> IO.succeedNow(true)).timeout(10.millis)
-        assertM(io.provideLayer(Clock.live))(isNone)
+        assertM(Live.live(io))(isNone)
+      },
+      testM("timeout repetition of uninterruptible effect") {
+        val effect = ZIO.unit.uninterruptible.forever
+
+        assertM(Live.live(effect.timeout(1.second)))(isNone)
       },
       testM("catchAllCause") {
         val io =

--- a/core-tests/shared/src/test/scala/zio/ZIOSpec.scala
+++ b/core-tests/shared/src/test/scala/zio/ZIOSpec.scala
@@ -10,7 +10,7 @@ import zio.duration._
 import zio.internal.Platform
 import zio.random.Random
 import zio.test.Assertion._
-import zio.test.TestAspect.{ flaky, forked, ignore, jvm, nonFlaky, scala2Only }
+import zio.test.TestAspect.{ flaky, forked, ignore, jvm, jvmOnly, nonFlaky, scala2Only }
 import zio.test._
 import zio.test.environment.{ Live, TestClock }
 
@@ -1773,7 +1773,7 @@ object ZIOSpec extends ZIOBaseSpec {
         val effect = ZIO.unit.uninterruptible.forever
 
         assertM(Live.live(effect.timeout(1.second)))(isNone)
-      },
+      } @@ jvmOnly,
       testM("catchAllCause") {
         val io =
           for {

--- a/core/shared/src/main/scala/zio/Fiber.scala
+++ b/core/shared/src/main/scala/zio/Fiber.scala
@@ -302,16 +302,10 @@ sealed trait Fiber[+E, +A] { self =>
 
       for {
         runtime <- ZIO.runtime[Any]
-        _       <- completeFuture.fork
+        _       <- completeFuture.forkDaemon // Cannot afford to NOT complete the promise, no matter what, so we fork daemon
       } yield new CancelableFuture[A](p.future) {
-        def cancel(): Future[Exit[Throwable, A]] = {
-          // For Dotty compatibility
-          val p = scala.concurrent.Promise[Exit[Throwable, A]]()
-          runtime.unsafeRunAsync(self.interrupt) { (exit: Exit[Nothing, Exit[E, A]]) =>
-            val _ = p.success(exit.flatten.mapError(f))
-          }
-          p.future
-        }
+        def cancel(): Future[Exit[Throwable, A]] =
+          runtime.unsafeRunToFuture[Nothing, Exit[Throwable, A]](self.interrupt.map(_.mapError(f)))
       }
     }.uninterruptible
 

--- a/core/shared/src/main/scala/zio/Runtime.scala
+++ b/core/shared/src/main/scala/zio/Runtime.scala
@@ -124,8 +124,8 @@ trait Runtime[+R] {
    *
    * This method is effectful and should only be used at the edges of your program.
    */
-  final def unsafeRunToFuture[E <: Throwable, A](io: ZIO[R, E, A]): CancelableFuture[A] =
-    unsafeRun(io.toFuture)
+  final def unsafeRunToFuture[E <: Throwable, A](zio: ZIO[R, E, A]): CancelableFuture[A] =
+    unsafeRun(zio.forkDaemon >>= (_.toFuture))
 
   /**
    * Constructs a new `Runtime` with the specified new environment.

--- a/core/shared/src/main/scala/zio/ZIO.scala
+++ b/core/shared/src/main/scala/zio/ZIO.scala
@@ -2555,37 +2555,28 @@ object ZIO extends ZIOCompanionPlatformSpecific {
    * [[scala.concurrent.ExecutionContext]] into a `ZIO`.
    */
   def fromFuture[A](make: ExecutionContext => scala.concurrent.Future[A]): Task[A] =
-    ZIO.uninterruptibleMask(restore =>
-      Task.descriptorWith { d =>
-        val ec = d.executor.asEC
-        ZIO.effect(make(ec)).flatMap {
-          f =>
-            val canceler: UIO[Unit] = f match {
-              case cancelable: CancelableFuture[A] =>
-                UIO.effectSuspendTotal(if (f.isCompleted) ZIO.unit else ZIO.fromFuture(_ => cancelable.cancel()).ignore)
-              case _ => ZIO.unit
-            }
-
-            val converted =
-              f.value
-                .fold(
-                  Task.effectAsyncInterrupt { (k: Task[A] => Unit) =>
-                    f.onComplete {
-                      case Success(a) => k(Task.succeedNow(a))
-                      case Failure(t) => k(Task.failNow(t))
-                    }(ec)
-
-                    Left(canceler)
-                  }
-                )(Task.fromTry(_))
-
-            (for {
-              fiber <- converted.fork
-              a     <- restore(fiber.join)
-            } yield a)
+    Task.descriptorWith { d =>
+      val ec = d.executor.asEC
+      ZIO.effect(make(ec)).flatMap { f =>
+        val canceler: UIO[Unit] = f match {
+          case cancelable: CancelableFuture[A] =>
+            UIO.effectSuspendTotal(if (f.isCompleted) ZIO.unit else ZIO.fromFuture(_ => cancelable.cancel()).ignore)
+          case _ => ZIO.unit
         }
+
+        f.value
+          .fold(
+            Task.effectAsyncInterrupt { (k: Task[A] => Unit) =>
+              f.onComplete {
+                case Success(a) => k(Task.succeedNow(a))
+                case Failure(t) => k(Task.failNow(t))
+              }(ec)
+
+              Left(canceler)
+            }
+          )(Task.fromTry(_))
       }
-    )
+    }
 
   /**
    * Imports a function that creates a [[scala.concurrent.Future]] from an

--- a/core/shared/src/main/scala/zio/ZIO.scala
+++ b/core/shared/src/main/scala/zio/ZIO.scala
@@ -2558,13 +2558,21 @@ object ZIO extends ZIOCompanionPlatformSpecific {
     Task.descriptorWith { d =>
       val ec = d.executor.asEC
       ZIO.effect(make(ec)).flatMap { f =>
+        val canceler: UIO[Unit] = f match {
+          case cancelable: CancelableFuture[A] =>
+            UIO.effectSuspendTotal(if (f.isCompleted) ZIO.unit else ZIO.fromFuture(_ => cancelable.cancel()).ignore)
+          case _ => ZIO.unit
+        }
+
         f.value
           .fold(
-            Task.effectAsync { (k: Task[A] => Unit) =>
+            Task.effectAsyncInterrupt { (k: Task[A] => Unit) =>
               f.onComplete {
                 case Success(a) => k(Task.succeedNow(a))
                 case Failure(t) => k(Task.failNow(t))
               }(ec)
+
+              Left(canceler)
             }
           )(Task.fromTry(_))
       }

--- a/core/shared/src/main/scala/zio/ZIO.scala
+++ b/core/shared/src/main/scala/zio/ZIO.scala
@@ -2555,28 +2555,37 @@ object ZIO extends ZIOCompanionPlatformSpecific {
    * [[scala.concurrent.ExecutionContext]] into a `ZIO`.
    */
   def fromFuture[A](make: ExecutionContext => scala.concurrent.Future[A]): Task[A] =
-    Task.descriptorWith { d =>
-      val ec = d.executor.asEC
-      ZIO.effect(make(ec)).flatMap { f =>
-        val canceler: UIO[Unit] = f match {
-          case cancelable: CancelableFuture[A] =>
-            UIO.effectSuspendTotal(if (f.isCompleted) ZIO.unit else ZIO.fromFuture(_ => cancelable.cancel()).ignore)
-          case _ => ZIO.unit
-        }
-
-        f.value
-          .fold(
-            Task.effectAsyncInterrupt { (k: Task[A] => Unit) =>
-              f.onComplete {
-                case Success(a) => k(Task.succeedNow(a))
-                case Failure(t) => k(Task.failNow(t))
-              }(ec)
-
-              Left(canceler)
+    ZIO.uninterruptibleMask(restore =>
+      Task.descriptorWith { d =>
+        val ec = d.executor.asEC
+        ZIO.effect(make(ec)).flatMap {
+          f =>
+            val canceler: UIO[Unit] = f match {
+              case cancelable: CancelableFuture[A] =>
+                UIO.effectSuspendTotal(if (f.isCompleted) ZIO.unit else ZIO.fromFuture(_ => cancelable.cancel()).ignore)
+              case _ => ZIO.unit
             }
-          )(Task.fromTry(_))
+
+            val converted =
+              f.value
+                .fold(
+                  Task.effectAsyncInterrupt { (k: Task[A] => Unit) =>
+                    f.onComplete {
+                      case Success(a) => k(Task.succeedNow(a))
+                      case Failure(t) => k(Task.failNow(t))
+                    }(ec)
+
+                    Left(canceler)
+                  }
+                )(Task.fromTry(_))
+
+            (for {
+              fiber <- converted.fork
+              a     <- restore(fiber.join)
+            } yield a)
+        }
       }
-    }
+    )
 
   /**
    * Imports a function that creates a [[scala.concurrent.Future]] from an
@@ -3239,7 +3248,7 @@ object ZIO extends ZIOCompanionPlatformSpecific {
 
   final class TimeoutTo[R, E, A, B](self: ZIO[R, E, A], b: B) {
     def apply[B1 >: B](f: A => B1)(duration: Duration): ZIO[R with Clock, E, B1] =
-      (self map f) raceFirst (ZIO.sleep(duration) as b)
+      (self map f) raceFirst (ZIO.sleep(duration) as b) // TODO: Make right-hand side interruptible?
   }
 
   final class BracketAcquire_[-R, +E](private val acquire: ZIO[R, E, Any]) extends AnyVal {

